Fixes #16986 - Add UI for template import/export

xprazak2 opened the pull request:
Adds UI for template import/export.

~TODO:~
* ~tests~
* ~linting~
* ~better error handling on export~

Import form fields - the asterisk for required fields with help needs adjusting:

![](https://i.imgur.com/cCpThex.png)

Export form fields:

![](https://i.imgur.com/Bqjv83l.png)

List of templates we tried to import:

![](https://i.imgur.com/e8Jt1OE.png)

xprazak2 commented:
Found a bug: when exporting, the settings for import are posted as well, will need to fix.

Rohoover commented:
@xprazak2 

How do you get to this page? I'm trying to understand if the preferred mechanism for importing or exporting should be a radio button (which I assume readjusts the fields). 

Can we separate out the "Use Default" button from the fields (they're touching).

Does the 'back' button take you back to the import/export form?

Does the pagination also exist on the bottom of the page as well?

ares commented:
I think it would be great to setup a short call where @xprazak2 could demo it, seeing it interactively would probably be better than just screenshots in this case. If you do that, please also send invitation to me :-)

sean797 commented on lib/foreman_templates/engine.rb:
```diff
@@ -28,6 +35,13 @@ class Engine < ::Rails::Engine
+             :caption => N_('Sync Templates'),
+             :parent => :hosts_menu,
+             :before => :ptables,
+             :turbolinks => false
```
I'm not sure I like another link on the Navigation? Have you considered adding a link on both Provisioning, Partition and Job template pages instead? 

I don't really support either option more, just a discussion point

xprazak2 replied:
That is a good idea, unfortunately the only way we can do it now is using Deface, which I would like to avoid. I'll gladly add them once we have proper extension points.

sean797 replied:
Since as I understand where waiting for theforeman/foreman#6499 & theforeman/foreman#6604 is it worth adding a plugin API for it now? I'm not sure if that's a big task or not..

sean797 commented on app/views/template_syncs/template_attrs.json.rabl:
```diff
@@ -0,0 +1,11 @@
+end
+
+node(:class_name) do |template|
+  template.class.name.underscore.split('_').map { |part| part.capitalize }.join(' ')
```
Can we change this, "Ptable" isn't as friendly to users as it could be?

xprazak2 replied:
Good point, changed.

xprazak2 commented:
updates in form:

* Add space between form buttons and fields
* Use icon for buttons and have description what the button does in tooltip
* remove the verbose setting, template diffs will be added as a follow-up

![sync-form](https://user-images.githubusercontent.com/2664090/55386214-4de22180-552f-11e9-86fa-79d20a6b5764.png)

updates in result:

* links for templates open in new tab
* breadcrumbs (not fully functional yet, changes in core needed to make them work with react-router)
* pagination moved down
* elipsis for file names

![sync-result](https://user-images.githubusercontent.com/2664090/55386221-4f134e80-552f-11e9-9b28-c147f507fb13.png)


I also added 'Not found' page that handles unknown routes

![not-found](https://user-images.githubusercontent.com/2664090/55386223-50dd1200-552f-11e9-9097-592d474f873e.png)

todos: 
* better error handling when sync goes wrong (https://github.com/theforeman/foreman/pull/6499 is needed)
* breadcrumbs fully functional (https://github.com/theforeman/foreman/pull/6604 or alternative is needed)
* title bar for list view

sean797 commented:
The new changes look awesome :+1:

terezanovotna commented:
following up our conversation @xprazak2  - here is an example of header for list 

<img width="1022" alt="Screen Shot 2019-04-03 at 22 03 13" src="https://user-images.githubusercontent.com/36040135/55509508-73b91480-565c-11e9-84a4-59fb9917b86f.png">


Few more visuals I have noticed: 

1)breadcrumbs text is a little bit off. [breadcrumbs](http://patternfly-react.surge.sh/patternfly-3/index.html?selectedKind=patternfly-react%2FNavigation%2FBreadcrumb&selectedStory=Breadcrumb&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs)

xprazak2 commented:
Adding a couple of commits that take care of the following:

* header for list view
* prop types
* linting
* nice errors

I had to remove breadcrumbs - we would like to have this for 1.22 and support for react-router together with breadcrumbs did not make it into core.

![](https://user-images.githubusercontent.com/2664090/55796193-a5e9cc80-5ac9-11e9-9f3d-3678379e32a1.png)

I think this is ready for review.

xprazak2 commented:
@theforeman/ui-ux can someone take a look please?

terezanovotna commented:
Looks good!

Can you make sure the font sizes in the list view are all same size? [example](https://rawgit.com/patternfly/patternfly-react/gh-pages/patternfly-3/index.html?selectedKind=patternfly-react%2FContent%20Views%2FList%20View&selectedStory=List%20of%20expandable%20items&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs)


and when we have breadcrumbs, will be even better
👍👍👍

ares commented:
Could we change scissors icon to the same we use in templates table? (check mark IIRC)
It would be great to have frontend code reviewed :-)

xprazak2 commented:
I adjusted the font size and replaced the icon:

![adjusted](https://user-images.githubusercontent.com/2664090/55952416-dd877e80-5c59-11e9-9dd1-52076a9d1a6d.png)

xprazak2 commented:
Updated npm package versions to be compatible with what we have in core. Update of redux-form will be needed. We will also have to add react-router and react-router-dom.

ares commented:
Meaning this is now blocked on core?

xprazak2 commented:
react-router and react-router-dom have been moved in [packaging](https://github.com/theforeman/foreman-packaging/commit/aae81dac30d118c35c999f3dea078ea657cc9684) and I opened https://github.com/theforeman/foreman/pull/6729

sharvit commented on webpack/components/NewTemplateSync/components/TextButtonField.js:
```diff
@@ -0,0 +1,212 @@
+  item: {},
+};
+
+export default TextButtonField;
```
This file contains too many things.
The UI team has a golden rule of one component per file. I believe following this rule here will make it much more readable and easier to maintain.

sharvit commented on webpack/components/NewTemplateSync/components/SyncSettingFields.js:
```diff
@@ -0,0 +1,70 @@
+
+  return syncType === 'import'
+    ? mapSettings(importSettings)
+    : mapSettings(exportSettings);
```
Could you make it more readable? I can imagine something like this:

```js
// in a helper file?
const addValidationToSetting = setting => {
  switch (setting.name) {
    case 'repo':
      return setting.merge({
        required: true,
        validate: [repoFormat(validationDataObj.repo)],
      });
    default:
      return setting;
  }
};

const settingsAry = syncType === 'import' ? importSettings : importSettings;

return (
  <React.Fragment>
    {settingsAry.map(setting => addValidationToSetting(setting)).map((setting, index) => (
      <SyncSettingField
        setting={setting}
        key={setting.name}
        disabled={disabled}
        resetField={resetField}
      />
    ))}
  </React.Fragment>
);
```

sharvit commented on webpack/components/NewTemplateSync/components/NewTemplateSyncForm/NewTemplateSyncForm.js:
```diff
@@ -0,0 +1,145 @@
+      { label: 'Export', value: 'export', permission: 'export' },
+    ];
+
+    this.state = {
```
Is there a way you can implement it without an internal state?
We have found out that where we mix redux and internal state things get messy and difficult to maintain.

xprazak2 replied:
Yes, there is a way. I even had that initially but it looked too complex for the task so I decided to just put it in state. I expect this will go away once we move to new forms framework.

Laviro commented on config/routes.rb:
```diff
@@ -1,6 +1,18 @@
+    end
+  end
+
+  match '/template_syncs/*page' => 'template_syncs#index', :via => [:get]
```
I know I am a bit late for the party,
but would you like to use react router now that it has been added ? :)

xprazak2 replied:
I would like that very much :), one problem is that we inject quite a lot of things into props when we `mount_react_component` in view which is sort of "context" for the plugin and I am not sure if I can make it work in a sane way with what we currently have in core. I will look into it though.

ares commented:
Could be a good hacking session. We decided to postpone merging this until plugins with react can be safely packaged.

xprazak2 commented:
Using @theforeman/vendor now and fixed comments except for 2.

sharvit commented on webpack/components/NewTemplateSync/NewTemplateSyncActions.js:
```diff
@@ -0,0 +1,32 @@
+
+export const getSyncSettings = url => dispatch => {
+  dispatch({ type: SYNC_SETTINGS_REQUEST });
+  return api
```
Would you mind migrating to `async/await` syntax?

It is recommended because we already started the work on the migration:
https://github.com/theforeman/foreman/pull/6901

It's include a new `eslint` rule that will block you from using `.then/.catch` and we will include it in the `@theforeman/env` package so it is better to migrate now so you won't get the errors later.

I would recommend using the eslint plugin `promise` as @amirfefer use in his PR.

sharvit commented on webpack/components/NewTemplateSync/components/NewTemplateSyncForm/NewTemplateSyncForm.js:
```diff
@@ -0,0 +1,145 @@
+import React from 'react';
+import { change } from 'redux-form';
```
In your opinion, can you finalize the work on the `formik` infrastructure and use it here once it is ready?
Is it going to drastically change the design?

What about refactoring into multiple PRs so we can gradually insert code which is ready?
- You can decide you only use it in the storybook.
- You can create a secret/experimental route (e.g. /ex/templates) and gradually deliver features to the new route until it is ready to replace the current route.

sharvit commented on webpack/components/NewTemplateSync/components/NewTemplateSyncForm/NewTemplateSyncFormConstants.js:
```diff
@@ -0,0 +1 @@
+export const formName = 'newTemplateSync';
```
The convention for such a constant is `FORM_NAME`, consider changing to `NEW_TEMPLATE_SYNC_FORM_NAME`.

jyejare commented:
Hello @xprazak2 ,

I tried testing this PR and could find following issues:

```
1. The metadata-export Mode option in export is missing the letter "t" in word "without" in tooltip. (**Low Severity**)
2. The reset button for export - negate is not working. (**Low Severity**)
3. The export to local dir is failing with - Error! Permission denied @ dir_s_mkdir - /usr/share/foreman/exported/provisioning_templates even though I set correct permissions. (**High/Urgent Severity**)
-> That is - owning to foreman-proxy and SELinux contest.
4. Error importing templates from git -  Warning! Error submitting data: 500 Internal Server Error (**High/Urgent Severity**)
```
As, Import and Export are both blocked I couldn't able to continue my testing. Please look into it.

ares commented:
@jyejare thanks for testing, I think your findings are not related to the UI (I hope at least), or do you confirm the API/Hammer works with the same paramters? I think it can be the environemnt/values provided to the form.

See https://theforeman.org/plugins/foreman_templates/6.0/index.html#7.1.3Permissiondenied@dir_s_mkdir-/some/provisioning_templates%E2%80%9D for more details

Based on the info you provided, the directory is owned by `foreman-proxy` user but it should be owned by `foreman` user

jyejare commented:
@ares I dont have foreman user in my setup. So shall I create the foreman self?

ares commented:
oh, if this is not production-like setup, then it needs to be owned by the user under which your Foreman instance run, `whoami` should give you the answer, it could be vagrant or jyejare :-)

xprazak2 commented:
I added fixes for the latest comments.

ares commented:
rubocop is unhappy

```
** Execute rubocop_foreman_templates
/home/jenkins/workspace/foreman_templates-pull-request/database/postgresql/label/fast/ruby/2.6/plugin/.rubocop.yml:88: `Metrics/PerceivedComplexity` is concealed by line 96
Error: The `Layout/IndentHash` cop has been renamed to `Layout/IndentFirstHashElement`.
(obsolete configuration found in /home/jenkins/workspace/foreman_templates-pull-request/database/postgresql/label/fast/ruby/2.6/plugin/.rubocop.yml, please update it)
RuboCop failed!
Running RuboCop...
Inspecting 21 files
```

ares commented:
I'm sorry, cop is now running but reporting many issues

```
Inspecting 21 files
CCCCCCCCCCCCCCCCCCCCC

Offenses:

/home/jenkins/workspace/foreman_templates-pull-request/database/postgresql/label/fast/ruby/2.5/plugin/app/controllers/api/v2/template_controller.rb:1:1: C: Style/FrozenStringLiteralComment: Missing magic comment # frozen_string_literal: true.
module Api
^
/home/jenkins/workspace/foreman_templates-pull-request/database/postgresql/label/fast/ruby/2.5/plugin/app/controllers/concerns/foreman/controller/parameters/template_params.rb:1:1: C: Style/FrozenStringLiteralComment: Missing magic comment # frozen_string_literal: true.
module Foreman
^
/home/jenkins/workspace/foreman_templates-pull-request/database/postgresql/label/fast/ruby/2.5/plugin/app/controllers/template_syncs_controller.rb:1:1: C: Style/FrozenStringLiteralComment: Missing magic comment # frozen_string_literal: true.
class TemplateSyncsController < ReactController
^
/home/jenkins/workspace/foreman_templates-pull-request/database/postgresql/label/fast/ruby/2.5/plugin/app/controllers/ui_template_syncs_controller.rb:1:1: C: Style/FrozenStringLiteralComment: Missing magic comment # frozen_string_literal: true.
require 'ostruct'
^
/home/jenkins/workspace/foreman_templates-pull-request/database/postgresql/label/fast/ruby/2.5/plugin/app/models/setting/template_sync.rb:1:1: C: Style/FrozenStringLiteralComment: Missing magic comment # frozen_string_literal: true.
class Setting
^
/home/jenkins/workspace/foreman_templates-pull-request/database/postgresql/label/fast/ruby/2.5/plugin/app/services/foreman_templates/cleaner.rb:1:1: C: Style/FrozenStringLiteralComment: Missing magic comment # frozen_string_literal: true.
module ForemanTemplates
^
/home/jenkins/workspace/foreman_templates-pull-request/database/postgresql/label/fast/ruby/2.5/plugin/app/services/foreman_templates/action.rb:1:1: C: Style/FrozenStringLiteralComment: Missing magic comment # frozen_string_literal: true.
module ForemanTemplates
^
/home/jenkins/workspace/foreman_templates-pull-request/database/postgresql/label/fast/ruby/2.5/plugin/app/services/foreman_templates/export_result.rb:1:1: C: Style/FrozenStringLiteralComment: Missing magic comment # frozen_string_literal: true.
module ForemanTemplates
^
/home/jenkins/workspace/foreman_templates-pull-request/database/postgresql/label/fast/ruby/2.5/plugin/app/services/foreman_templates/parse_result.rb:1:1: C: Style/FrozenStringLiteralComment: Missing magic comment # frozen_string_literal: true.
module ForemanTemplates
^
/home/jenkins/workspace/foreman_templates-pull-request/database/postgresql/label/fast/ruby/2.5/plugin/app/services/foreman_templates/path_access_exception.rb:1:1: C: Style/FrozenStringLiteralComment: Missing magic comment # frozen_string_literal: true.
module ForemanTemplates
^
/home/jenkins/workspace/foreman_templates-pull-request/database/postgresql/label/fast/ruby/2.5/plugin/app/services/foreman_templates/template_exporter.rb:1:1: C: Style/FrozenStringLiteralComment: Missing magic comment # frozen_string_literal: true.
require 'shellwords'
^
/home/jenkins/workspace/foreman_templates-pull-request/database/postgresql/label/fast/ruby/2.5/plugin/app/services/foreman_templates/template_importer.rb:1:1: C: Style/FrozenStringLiteralComment: Missing magic comment # frozen_string_literal: true.
module ForemanTemplates
^
/home/jenkins/workspace/foreman_templates-pull-request/database/postgresql/label/fast/ruby/2.5/plugin/app/helpers/foreman_templates_helper.rb:1:1: C: Style/FrozenStringLiteralComment: Missing magic comment # frozen_string_literal: true.
module ForemanTemplatesHelper
^
/home/jenkins/workspace/foreman_templates-pull-request/database/postgresql/label/fast/ruby/2.5/plugin/lib/foreman_templates.rb:1:1: C: Style/FrozenStringLiteralComment: Missing magic comment # frozen_string_literal: true.
require 'foreman_templates/engine'
^
/home/jenkins/workspace/foreman_templates-pull-request/database/postgresql/label/fast/ruby/2.5/plugin/lib/foreman_templates/version.rb:1:1: C: Style/FrozenStringLiteralComment: Missing magic comment # frozen_string_literal: true.
module ForemanTemplates
^
/home/jenkins/workspace/foreman_templates-pull-request/database/postgresql/label/fast/ruby/2.5/plugin/lib/foreman_templates/engine.rb:1:1: C: Style/FrozenStringLiteralComment: Missing magic comment # frozen_string_literal: true.
require 'fileutils'
^
/home/jenkins/workspace/foreman_templates-pull-request/database/postgresql/label/fast/ruby/2.5/plugin/test/functional/api/v2/template_controller_test.rb:1:1: C: Style/FrozenStringLiteralComment: Missing magic comment # frozen_string_literal: true.
require 'test_plugin_helper'
^
/home/jenkins/workspace/foreman_templates-pull-request/database/postgresql/label/fast/ruby/2.5/plugin/test/test_plugin_helper.rb:1:1: C: Style/FrozenStringLiteralComment: Missing magic comment # frozen_string_literal: true.
# Code coverage
^
/home/jenkins/workspace/foreman_templates-pull-request/database/postgresql/label/fast/ruby/2.5/plugin/test/unit/template_exporter_test.rb:1:1: C: Style/FrozenStringLiteralComment: Missing magic comment # frozen_string_literal: true.
require 'test_plugin_helper'
^
/home/jenkins/workspace/foreman_templates-pull-request/database/postgresql/label/fast/ruby/2.5/plugin/test/unit/action_test.rb:1:1: C: Style/FrozenStringLiteralComment: Missing magic comment # frozen_string_literal: true.
require 'test_plugin_helper'
^
/home/jenkins/workspace/foreman_templates-pull-request/database/postgresql/label/fast/ruby/2.5/plugin/test/unit/template_importer_test.rb:1:1: C: Style/FrozenStringLiteralComment: Missing magic comment # frozen_string_literal: true.
require 'test_plugin_helper'
^

```

ares commented:
Thanks @xprazak2, I'm finally merging this! I opened a tracker of details that shouldn't block this PR anymore - https://projects.theforeman.org/issues/27961
